rules/DowngradePhp70/Rector/Instanceof_/DowngradeInstanceofThrowableRector.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\DowngradePhp70\Rector\Instanceof_;
+
+use PhpParser\Node;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr\BinaryOp\BooleanOr;
+use PhpParser\Node\Expr\Instanceof_;
+use PhpParser\Node\Name;
+use PhpParser\Node\Name\FullyQualified;
+use Rector\Core\Rector\AbstractRector;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * Throwable was introduced in PHP 7.0 so to support older versions we need to also check for Exception.
+ * @see https://www.php.net/manual/en/class.throwable.php
+ * @see \Rector\Tests\DowngradePhp70\Rector\Instanceof_\DowngradeInstanceofThrowableRector\DowngradeInstanceofThrowableRectorTest
+ */
+final class DowngradeInstanceofThrowableRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add `instanceof Exception` check as a fallback to `instanceof Throwable` to support exception hierarchies in PHP 5', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+return $e instanceof \Throwable;
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+return $e instanceof \Throwable || $e instanceof \Exception;
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [Instanceof_::class];
+    }
+
+    /**
+     * @param Instanceof_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! ($node->class instanceof Name && $node->class->toCodeString() === '\\Throwable')) {
+            return null;
+        }
+
+        $exceptionFallback = new Instanceof_($node->expr, new FullyQualified('Exception'));
+
+        /** @var Node $parentNode */
+        $parentNode = $node->getAttribute(AttributeKey::PARENT_NODE);
+        if ($parentNode instanceof BooleanOr && $parentNode->left === $node && json_encode($parentNode->right) === json_encode($exceptionFallback)) {
```

Comment:
To avoid `InfiniteLoopTraversingException`, I need to add a termination condition. However `json_encode` calls just return `false`. Is there some working way to check if two `Expr` nodes are structurally identical?

Reply:
Or maybe I could just annotate already processed node somehow?

Reply:
You can exclude rule if it not make endless loop:

https://github.com/rectorphp/rector-src/blob/34c61ba4498150972db843fca807416540ca71e8/src/Validation/InfiniteLoopValidator.php#L24

You can compare node with NodeComparator service

Reply:
It will make an endless loop. I found an attribute based approach here: https://github.com/rectorphp/rector-src/blob/fc10fce13dcf9767f54e4202b509020fed338645/rules/PhpSpecToPHPUnit/Rector/Class_/PhpSpecClassToPHPUnitClassRector.php#L63-L66